// twitch-webhooks/es/WebHookListener.mjs
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
import Logger from '@d-fischer/logger';
import getRawBody from '@d-fischer/raw-body';
import { Server } from 'httpanda';
import { extractUserId } from 'twitch';
import LegacyAdapter from "./Adapters/LegacyAdapter.mjs";
import BanEventSubscription from "./Subscriptions/BanEventSubscription.mjs";
import ExtensionTransactionSubscription from "./Subscriptions/ExtensionTransactionSubscription.mjs";
import FollowsFromUserSubscription from "./Subscriptions/FollowsFromUserSubscription.mjs";
import FollowsToUserSubscription from "./Subscriptions/FollowsToUserSubscription.mjs";
import ModeratorEventSubscription from "./Subscriptions/ModeratorEventSubscription.mjs";
import StreamChangeSubscription from "./Subscriptions/StreamChangeSubscription.mjs";
import SubscriptionEventSubscription from "./Subscriptions/SubscriptionEventSubscription.mjs";
import UserChangeSubscription from "./Subscriptions/UserChangeSubscription.mjs";
/**
 * A WebHook listener you can track changes in various channel and user data with.
 */
var WebHookListener = /** @class */ (function () {
    /**
     * Creates a new WebHook listener.
     *
     * @param twitchClient The TwitchClient instance to use for user info and API requests.
     * @param adapter The connection adapter.
     * @param config
     */
    function WebHookListener(twitchClient, adapter, config) {
        if (config === void 0) { config = {}; }
        var _a;
        this._subscriptions = new Map();
        this._twitchClient = twitchClient;
        this._adapter = adapter;
        this._hookValidity = config.hookValidity;
        this._logger = new Logger(__assign({ name: 'twitch-webhooks', emoji: true }, (_a = config.logger, (_a !== null && _a !== void 0 ? _a : {}))));
    }
    /**
     * Creates a new WebHook listener.
     *
     * @deprecated Use the normal constructor instead.
     *
     * @param twitchClient The TwitchClient instance to use for user info and API requests.
     * @param config
     */
    WebHookListener.create = function (twitchClient, config) {
        if (config === void 0) { config = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var adapter;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, LegacyAdapter.create(config)];
                    case 1:
                        adapter = _a.sent();
                        return [2 /*return*/, new WebHookListener(twitchClient, adapter, config)];
                }
            });
        });
    };
    /**
     * Starts the backing server and listens to incoming WebHook notifications.
     */
    WebHookListener.prototype.listen = function () {
        return __awaiter(this, void 0, void 0, function () {
            var server, listenerPort;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this._server) {
                            throw new Error('Trying to listen while already listening');
                        }
                        server = this._adapter.createHttpServer();
                        this._server = new Server({
                            server: server,
                            // eslint-disable-next-line @typescript-eslint/no-explicit-any
                            onError: function (e, req) {
                                if (e.code === 404) {
                                    _this._logger.warn("Access to unknown URL/method attempted: " + req.method + " " + req.url);
                                }
                            }
                        });
                        // needs to be first in chain but run last, for proper logging of status
                        this._server.use(function (req, res, next) {
                            setImmediate(function () {
                                _this._logger.debug(req.method + " " + req.path + " - " + res.statusCode);
                            });
                            next();
                        });
                        this._server.all('/:id', this._createHandleRequest());
                        return [4 /*yield*/, this._adapter.getListenerPort()];
                    case 1:
                        listenerPort = _a.sent();
                        return [4 /*yield*/, this._server.listen(listenerPort)];
                    case 2:
                        _a.sent();
                        this._logger.info("Listening on port " + listenerPort);
                        return [4 /*yield*/, Promise.all(__spread(this._subscriptions.values()).map(function (sub) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                return [2 /*return*/, sub.start()];
                            }); }); }))];
                    case 3:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Stops the backing server, suspending all active subscriptions.
     */
    WebHookListener.prototype.unlisten = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this._server) {
                            throw new Error('Trying to unlisten while not listening');
                        }
                        return [4 /*yield*/, this._server.close()];
                    case 1:
                        _a.sent();
                        this._server = undefined;
                        return [4 /*yield*/, Promise.all(__spread(this._subscriptions.values()).map(function (sub) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                return [2 /*return*/, sub.suspend()];
                            }); }); }))];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Applies middleware that handles WebHooks to a connect-compatible app (like express).
     *
     * @param app The app the middleware should be applied to.
     */
    WebHookListener.prototype.applyMiddleware = function (app) {
        var pathPrefix = this._adapter.pathPrefix;
        if (pathPrefix) {
            pathPrefix = "/" + pathPrefix.replace(/^\/|\/$/, '');
        }
        var paramParser = function (req, res, next) {
            var _a = __read(req.path.split('/'), 2), id = _a[1];
            req.param = req.params = { id: id };
            next();
        };
        var requestHandler = this._createHandleRequest();
        if (pathPrefix) {
            app.use(pathPrefix, paramParser, requestHandler);
        }
        else {
            app.use(paramParser, requestHandler);
        }
    };
    /**
     * Subscribes to events representing a user changing a public setting or their email address.
     *
     * @param user The user for which to get notifications about changing a setting.
     * @param handler The function that will be called for any new notifications.
     * @param withEmail Whether to subscribe to email address changes. This requires an additional scope (user:read:email).
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToUserChanges = function (user, handler, withEmail, validityInSeconds) {
        if (withEmail === void 0) { withEmail = false; }
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var userId, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        userId = extractUserId(user);
                        subscription = new UserChangeSubscription(userId, handler, withEmail, this, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /**
     * Subscribes to events representing a user being followed by other users.
     *
     * @param user The user for which to get notifications about the users they will be followed by.
     * @param handler The function that will be called for any new notifications.
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToFollowsToUser = function (user, handler, validityInSeconds) {
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var userId, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        userId = extractUserId(user);
                        subscription = new FollowsToUserSubscription(userId, handler, this, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /**
     * Subscribes to events representing a user following other users.
     *
     * @param user The user for which to get notifications about the users they will follow.
     * @param handler The function that will be called for any new notifications.
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToFollowsFromUser = function (user, handler, validityInSeconds) {
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var userId, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        userId = extractUserId(user);
                        subscription = new FollowsFromUserSubscription(userId, handler, this, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /**
     * Subscribes to events representing a stream changing, i.e. going live, offline or changing its title or category.
     *
     * @param user The user for which to get notifications about their streams changing.
     * @param handler The function that will be called for any new notifications.
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToStreamChanges = function (user, handler, validityInSeconds) {
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var userId, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        userId = extractUserId(user);
                        subscription = new StreamChangeSubscription(userId, handler, this, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /**
     * Subscribes to events representing the start or end of a channel subscription.
     *
     * @param user The user for which to get notifications about subscriptions to their channel.
     * @param handler The function that will be called for any new notifications.
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToSubscriptionEvents = function (user, handler, validityInSeconds) {
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var userId, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        userId = extractUserId(user);
                        subscription = new SubscriptionEventSubscription(userId, handler, this, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /**
     * Subscribes to events representing a ban or unban.
     *
     * @param broadcaster The broadcaster for which to get notifications about bans or unbans in their channel.
     * @param handler The function that will be called for any new notifications.
     * @param user The user that events will be sent for. If not given, events will be sent for all users.
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToBanEvents = function (broadcaster, handler, user, validityInSeconds) {
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var broadcasterId, userId, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        broadcasterId = extractUserId(broadcaster);
                        userId = user ? extractUserId(user) : undefined;
                        subscription = new BanEventSubscription(broadcasterId, handler, this, userId, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /**
     * Subscribes to events representing a user gaining or losing moderator privileges in a channel.
     *
     * @param broadcaster The broadcaster for which to get notifications about moderator changes in their channel.
     * @param handler The function that will be called for any new notifications.
     * @param user The user that events will be sent for. If not given, events will be sent for all users.
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToModeratorEvents = function (broadcaster, handler, user, validityInSeconds) {
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var broadcasterId, userId, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        broadcasterId = extractUserId(broadcaster);
                        userId = user ? extractUserId(user) : undefined;
                        subscription = new ModeratorEventSubscription(broadcasterId, handler, this, userId, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /**
     * Subscribes to extension transactions.
     *
     * @param extensionId The extension ID for which to get notifications about transactions.
     * @param handler The function that will be called for any new notifications.
     * @param validityInSeconds The validity of the WebHook, in seconds.
     *
     * Please note that this doesn't mean that you don't get any notifications after the given time. The hook will be automatically refreshed.
     *
     * This is meant for debugging issues. Please don't set it unless you know what you're doing.
     */
    WebHookListener.prototype.subscribeToExtensionTransactions = function (extensionId, handler, validityInSeconds) {
        if (validityInSeconds === void 0) { validityInSeconds = this._hookValidity; }
        return __awaiter(this, void 0, void 0, function () {
            var subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        subscription = new ExtensionTransactionSubscription(extensionId, handler, this, validityInSeconds);
                        return [4 /*yield*/, subscription.start()];
                    case 1:
                        _a.sent();
                        this._subscriptions.set(subscription.id, subscription);
                        return [2 /*return*/, subscription];
                }
            });
        });
    };
    /** @private */
    WebHookListener.prototype._buildHookUrl = function (id) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var protocol, hostName, externalPort, protocolDefaultPort, hostPortion, pathPrefix;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        protocol = this._adapter.connectUsingSsl ? 'https' : 'http';
                        return [4 /*yield*/, this._adapter.getHostName()];
                    case 1:
                        hostName = _b.sent();
                        return [4 /*yield*/, this._adapter.getExternalPort()];
                    case 2:
                        externalPort = _b.sent();
                        protocolDefaultPort = this._adapter.connectUsingSsl ? 443 : 80;
                        hostPortion = externalPort === protocolDefaultPort ? hostName : hostName + ":" + externalPort;
                        pathPrefix = (_a = this._adapter.pathPrefix) === null || _a === void 0 ? void 0 : _a.replace(/^\/|\/$/, '');
                        return [2 /*return*/, protocol + "://" + hostPortion + (pathPrefix ? '/' : '') + (pathPrefix !== null && pathPrefix !== void 0 ? pathPrefix : '') + "/" + id];
                }
            });
        });
    };
    /** @private */
    WebHookListener.prototype._changeIdOfSubscription = function (oldId, newId) {
        var sub = this._subscriptions.get(oldId);
        if (sub) {
            this._subscriptions.delete(oldId);
            this._subscriptions.set(newId, sub);
        }
    };
    /** @private */
    WebHookListener.prototype._dropSubscription = function (id) {
        this._subscriptions.delete(id);
    };
    WebHookListener.prototype._createHandleRequest = function () {
        var _this = this;
        return function (req, res, next) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(req.method === 'GET')) return [3 /*break*/, 1];
                        this._handleVerification(req, res);
                        return [3 /*break*/, 3];
                    case 1:
                        if (!(req.method === 'POST')) return [3 /*break*/, 3];
                        return [4 /*yield*/, this._handleNotification(req, res)];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3:
                        next();
                        return [2 /*return*/];
                }
            });
        }); };
    };
    WebHookListener.prototype._handleVerification = function (req, res) {
        var _a;
        var id = req.param.id;
        var subscription = this._subscriptions.get(id);
        if (subscription) {
            var hubMode = (_a = req.query) === null || _a === void 0 ? void 0 : _a['hub.mode'];
            if (hubMode === 'subscribe') {
                subscription._verify();
                res.writeHead(202);
                res.end(req.query['hub.challenge']);
                this._logger.debug("Successfully subscribed to hook: " + id);
            }
            else if (hubMode === 'unsubscribe') {
                this._subscriptions.delete(id);
                res.writeHead(200);
                res.end();
                this._logger.debug("Successfully unsubscribed from hook: " + id);
            }
            else if (hubMode === 'denied') {
                this._logger.error("Subscription denied to hook: " + id + " (" + req.query['hub.reason'] + ")");
                res.writeHead(200);
                res.end();
            }
            else {
                this._logger.warn("Unknown hub.mode " + hubMode + " for hook: " + id);
                res.writeHead(400);
                res.end();
            }
        }
        else {
            this._logger.warn("Verification of unknown hook attempted: " + id);
            res.writeHead(410);
            res.end();
        }
    };
    WebHookListener.prototype._handleNotification = function (req, res) {
        return __awaiter(this, void 0, void 0, function () {
            var body, id, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, getRawBody(req, true)];
                    case 1:
                        body = _a.sent();
                        id = req.param.id;
                        subscription = this._subscriptions.get(id);
                        if (subscription) {
                            res.writeHead(202);
                            res.end();
                            if (subscription._handleData(body, req.headers['x-hub-signature'])) {
                                this._logger.debug("Successfully verified notification signature for hook: " + id);
                            }
                            else {
                                this._logger.warn("Failed to verify notification signature for hook: " + id + ". " +
                                    'This might be caused by Twitch still sending notifications with an old secret and is perfectly normal a few times after you just restarted the script.\n' +
                                    'If the problem persists over a long period of time, please file an issue.');
                            }
                        }
                        else {
                            this._logger.warn("Notification for unknown hook received: " + id);
                            res.writeHead(410);
                            res.end();
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    return WebHookListener;
}());
export default WebHookListener;
